### PR TITLE
[lit-next] Adds UpdatingController

### PR DESCRIPTION
**Pull request by sorvell**
Adds `UpdatingController`, a class which can interact with an `UpdatingElement` by hooking into its lifecycle, including connected, disconnected, update, updated, and requestUpdate.

### CODE REVIEW

**Comment by daKmoR** [comment]
this is going to be soooo useful 🤗

thaaankks 🙇‍♂️

**Review by kevinpschaaf — CHANGES_REQUESTED** [review]

File: packages/labs/src/controllers/updating-controller.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+   * updates.
+   * @param changedProperties
+   */
+  update(_changedProperties: PropertyValues) {}
+
+  /**
+   * Runs after the controller's element updates after its `updated` method.
+   * @param changedProperties
+   */
+  updated(_changedProperties: PropertyValues) {}
```
[the remark this answers is not in the document]
Comment (sorvell) [reply packages/labs/src/controllers/updating-controller.ts:78]: Following the convention used in UpdatingElement where you can safely call the `super` version (even though you don't have to).

File: packages/labs/docs/CONTROLLERS.md [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+
+Can be used to add global event listeners, or fire events for a context API.
+
+### disconnectedCallback
+
+Can be use to clean up work done in connected.
+
+### willUpdate
+
+Can be used to calculate derived state using element state. Note, this will be called during server rendering and therefore should not access element DOM.
```
[the remark this answers is not in the document]
Comment (sorvell) [reply packages/labs/docs/CONTROLLERS.md:107]: Tracked via https://github.com/Polymer/lit-html/issues/1417

File: packages/labs/src/index.ts [hunk 3]
```diff
@@ -11,5 +11,4 @@
  * subject to an additional IP rights grant found at
  * http://polymer.github.io/PATENTS.txt
  */
-export {};
-console.log("there's nothing here");
+export * from './controllers/updating-controller.js';
```
[the remark this answers is not in the document]
Comment (sorvell) [reply packages/labs/src/index.ts:14]: Done.

File: packages/benchmarks/lit-element/list/index.ts [hunk 4]
```diff
@@ -15,6 +15,19 @@ import {html, LitElement, css, PropertyDeclaration} from 'lit-element';
...
+    let resolve: () => void;
+    const p = new Promise((r) => (resolve = r));
+    document.addEventListener('readystatechange', async () => {
+      if (document.readyState === 'complete') {
+        resolve();
+      }
+    });
+    await p;
+  }
+  await new Promise((r) => setTimeout(r));
```
[the remark this answers is not in the document]
Comment (sorvell) [reply packages/benchmarks/lit-element/list/index.ts]: Done

File: packages/labs/docs/CONTROLLERS.md [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+- `addController(controller)`: adds a controller to the element's lifecycle. Controllers should typically be added in an element's constructor so they can access the entire lifecycle of the element.
+
+The implementation of an example controller would look something like:
+
+```js
+class MyController {
+  host: UpdatingElement;
+
+  #value;
+  _interval;
```
[the remark this answers is not in the document]
Comment (sorvell) [reply packages/labs/docs/CONTROLLERS.md:60]: Done

File: packages/labs/docs/CONTROLLERS.md [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+
+The controller can have basically any set of state, methods, and external subscriptions. It's just an object that interacts with its host element's state and lifecycle APIs.
+
+Note, controllers can also be composed inside other objects like lit-html directives.
+
+## Lifecycle use cases
+
+### constructor
+
+Classes have a constructor which can be used as usual for initialization. Controllers can be factory functions as well.
```
[the remark this answers is not in the document]
Comment (sorvell) [reply packages/labs/docs/CONTROLLERS.md:95]: Clarified.

File: packages/labs/docs/CONTROLLERS.md [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+
+### connectedCallback
+
+Can be used to add global event listeners, or fire events for a context API.
```
[the remark this answers is not in the document]
Comment (sorvell) [reply packages/labs/docs/CONTROLLERS.md:99]: Clarified.

File: packages/labs/docs/CONTROLLERS.md [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+  _interval;
+
+  get value() {
+    return this.#value;
+  }
+
+  set value(v) {
+    this.#value = v;
+    this.host.requestUpdate();
+  }
```
[the remark this answers is not in the document]
Comment (sorvell) [reply packages/labs/docs/CONTROLLERS.md:69]: Added section in doc about exploring an `@property` feature.

**Comment by sorvell** [comment]
Closing in favor of https://github.com/Polymer/lit-html/pull/1435

**Pull request closed without merging**